pkg/status/bird_status.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/olekukonko/tablewriter"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// birdStatus is a structure containing details about bird.
+type birdStatus struct {
+	ready            bool
+	version          string
+	routeID          string
+	serverTime       string
+	lastBootTime     string
+	lastReconfigTime string
+}
+
+func (b *birdStatus) toNodeStatusAPI() apiv3.CalicoNodeBirdStatus {
+	return apiv3.CalicoNodeBirdStatus{
+		Ready:            b.ready,
+		Version:          b.version,
+		RouteID:          b.routeID,
+		ServerTime:       b.serverTime,
+		LastBootTime:     b.lastBootTime,
+		LastReconfigTime: b.lastReconfigTime,
+	}
+}
+
+// Unmarshal from a line in the BIRD protocol output.  Returns true if
+// successful, false otherwise.
+func (b *birdStatus) unmarshalBIRD(line string) bool {
+	// Peer names will be of the format described by bgpPeerRegex.
+	log.Debugf("Parsing line: %s", line)
+
+	if strings.Contains(line, "BIRD") && strings.Contains(line, "ready") {
```

Comment:
We should make sure to have extensive tests in case BIRD every updates / changes its output format, since this logic is highly dependent on that.

Reply:
Agree. I'm thinking to run tests with Node ST on kind. Couple of thoughts now.
**Infra**
- v4 cluster 
- Dual stack cluster
- Dual tor cluster

**BGP Peer type**
- NodeMesh
- BGPPeer

**Possible test cases**
- Node restart
- external BGP peer restart
- Pod restart (trigger a kernel route update)
- Kill bird or bird6 processes

Reply:
Could also use policy to block TCP port 179 on one or more nodes to simulate network failure?

Reply:
I'm sure they could be configured away for this test :)

Reply:
If not, perhaps `tc qdisc add dev eth0 root netem loss 100%`.